File: scripts/check_nav_v2_legacy_quality_cleanup_decision_v1.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
CONFIG = ROOT / "config/nav-v2-legacy-quality-cleanup-decision-v1.json"
SQL = ROOT / "supabase/prototypes/nav_v2_legacy_quality_cleanup_plan_v1.sql"
MIGRATIONS = ROOT / "supabase/migrations"


def require(condition: bool, message: str) -> None:
    if not condition:
        raise SystemExit(f"FAIL: {message}")


def main() -> None:
    config = json.loads(CONFIG.read_text(encoding="utf-8"))
    sql = SQL.read_text(encoding="utf-8")
    lower = sql.lower()

    require(config["status"] == "repository_only_decision_package", "status escaped decision package")
    require(config["production_applied"] is False, "production_applied must remain false")
    require(config["production_ready"] is False, "production_ready must remain false")
    require(config["writes_allowed"] is False, "writes_allowed must remain false")
    require(config["selected_option"] is None, "owner option was selected automatically")
    require(config["inventory_snapshot"]["total_open_rows"] == 46, "inventory total changed")
    require(config["inventory_snapshot"]["obsolete_privacy_conflict"] == 40, "privacy-conflict total changed")
    require(len(config["owner_options"]) == 3, "owner option inventory changed")
    require(config["owner_options"][0]["id"] == "gradual_on_touch", "safest option is no longer first")

    required = [
        "nav_v2_classify_legacy_quality_task_v1",
        "nav_v2_plan_legacy_quality_cleanup_v1",
        "'writes_performed', false",
        "'production_ready', false",
        "'selected_option', null",
        "obsolete_privacy_conflict",
        "replace_object_context",
        "replace_representation",
        "manual_review",
        "owner_options",
        "mandatory_stops",
    ]
    for marker in required:
        require(marker in sql, f"missing SQL marker: {marker}")

    forbidden_dml = ["insert into public.", "update public.", "delete from public.", "truncate public."]
    for marker in forbidden_dml:
        require(marker not in lower, f"planner contains business DML: {marker}")

    # Legacy source identifiers contain seller_name/buyer_name by design. Block only
    # actual column/JSON-key dependencies, not the source strings being classified.
    forbidden_field_patterns = [
        r"\b(?:d|t)\.(?:seller_name|buyer_name|seller_phone|buyer_phone|email|passport|snils|inn)\b",
        r"->>\s*'(?:seller_name|buyer_name|seller_phone|buyer_phone|email|passport|snils|inn)'",
        r"#>>\s*'\{[^}]*?(?:seller_name|buyer_name|seller_phone|buyer_phone|email|passport|snils|inn)[^}]*\}'",
    ]
    for pattern in forbidden_field_patterns:
        require(re.search(pattern, lower) is None, f"planner depends on PII field pattern: {pattern}")

    require("assigned_to" not in lower and "created_by" not in lower, "planner exposes employee assignment data")
    require("score" not in lower and "performance" not in lower, "planner contains employee evaluation semantics")
    require("grant execute" in lower and "to service_role" in lower, "service-only execute contract missing")

    leaked = [path.name for path in MIGRATIONS.glob("*legacy*quality*cleanup*")]
    require(not leaked, f"cleanup planner leaked into migrations: {leaked}")

    print("Navigator v2 legacy quality cleanup decision source contract passed")


if __name__ == "__main__":
    main()

File: scripts/check-nav-v2-legacy-quality-cleanup-decision-v1.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const config = JSON.parse(fs.readFileSync(path.join(root, 'config/nav-v2-legacy-quality-cleanup-decision-v1.json'), 'utf8'));
const sql = fs.readFileSync(path.join(root, 'supabase/prototypes/nav_v2_legacy_quality_cleanup_plan_v1.sql'), 'utf8');

function assert(condition, message) {
  if (!condition) throw new Error(message);
}

assert(config.contract_version === 1, 'unexpected cleanup contract version');
assert(config.status === 'repository_only_decision_package', 'cleanup contract escaped repository-only status');
assert(config.production_ready === false, 'cleanup contract claims production readiness');
assert(config.writes_allowed === false, 'cleanup contract allows writes');
assert(config.selected_option === null, 'cleanup option was selected without owner decision');
assert(config.inventory_snapshot.total_open_rows === 46, 'cleanup inventory differs from 46');
assert(config.inventory_snapshot.obsolete_privacy_conflict === 40, 'privacy-conflict inventory differs from 40');
assert(config.owner_options.length === 3, 'owner options must remain three');
assert(config.owner_options.filter((item) => item.recommended).length === 1, 'exactly one safest option must be marked');

for (const source of config.managed_legacy_sources) assert(sql.includes(source), `missing legacy source ${source}`);
for (const classification of config.classifications) assert(sql.includes(classification), `missing classification ${classification}`);
assert(sql.includes("'selected_option', null"), 'planner selected an option');
assert(sql.includes("'writes_performed', false"), 'planner write evidence missing');
assert(sql.includes("'production_ready', false"), 'planner production gate missing');
assert(sql.includes('order by source, deal_id, task_id'), 'planner output is not deterministically ordered');
assert(!/\b(insert|update|delete|truncate)\s+(into\s+|from\s+)?public\./i.test(sql), 'planner contains public-table DML');
assert(!/(seller_name|buyer_name|seller_phone|buyer_phone|passport|snils|\binn\b)/i.test(sql), 'planner contains PII dependency');
assert(!/(assigned_to|created_by|employee_score|performance)/i.test(sql), 'planner contains employee evaluation or assignment data');
assert(sql.includes('to service_role'), 'service-only grant is missing');

console.log('Navigator v2 legacy quality cleanup decision semantic contract passed');
